// src/spectralmc/proto/simulation.proto
syntax = "proto3";

package spectralmc.proto;

import "common.proto";

// ──────────────────────────── Simulation Parameters ───────────────────────
message SimulationParamsProto {
  uint32 skip = 1;
  uint32 timesteps = 2;
  uint32 network_size = 3;
  uint32 batches_per_mc_run = 4;
  uint32 threads_per_block = 5;
  uint32 mc_seed = 6;
  uint32 buffer_size = 7;
  PrecisionProto dtype = 8;
}

// ──────────────────────────── Black-Scholes Config ────────────────────────
message BlackScholesConfigProto {
  SimulationParamsProto sim_params = 1;
  bool simulate_log_return = 2;
  bool normalize_forwards = 3;
}

// ──────────────────────────── Option Contract ─────────────────────────────
message OptionContractProto {
  double X0 = 1;  // Spot price
  double K = 2;   // Strike price
  double T = 3;   // Time to maturity
  double r = 4;   // Risk-free rate
  double d = 5;   // Dividend yield
  double v = 6;   // Volatility
}

// ──────────────────────────── Pricing Results ─────────────────────────────
message PricingResultsProto {
  double call_price = 1;
  double put_price = 2;
  double underlying = 3;
  double forward = 4;
  double gamma = 5;
}

// ──────────────────────────── Sobol Config ────────────────────────────────
message BoundSpecProto {
  double lower = 1;
  double upper = 2;
}

message SobolConfigProto {
  uint32 dimensions = 1;
  uint32 seed = 2;
  map<string, BoundSpecProto> domain_bounds = 3;
}
